File: 2021/3/py/main-a.py
import argparse

def solve(raw_input):
  input = raw_input.split()
  gamma = ''
  for j in range(0, len(input[0])):
    count = 0
    for i in range(0, len(input)):
      count+= (input[i][j] == '1') - (input[i][j] == '0')

    gamma += '0' * (count < 0) + '1' * (count >= 0)

  not_gamma = "".join([str(1 - int(x)) for x in gamma])
  power = int(gamma,2) * int(not_gamma,2)

  print(f"{gamma=}")
  print(f"{not_gamma=}")
  print(f"{power=}")


def main(filename):
  # Read file and convert to ints
  input = []
  with open(filename) as file:
    input = file.read()
    solve(input)

if __name__ == '__main__':

  # Simple argparse for the input
  parser = argparse.ArgumentParser()
  parser.add_argument("filename", type=str, nargs="?", default="input.txt",
        help="Enter the name of the file (default = input.txt)")
  parser.add_argument("-i", "--interactive", default="false", action="store_true", dest="interactive",
        help="If flag is set, the input is interactive")
  args = parser.parse_args()

  # Run the program
  main(args.filename)

File: 2021/3/cpp/main-a.cpp
#include <iostream>
#include <fstream>
#include <string_view>
#include <string>
#include <filesystem>
namespace fs = std::filesystem;

std::vector<std::string> split(const std::string& s, const std::string& delim = " ")
{
  std::vector<std::string> tokens;

  size_t start = 0;
  size_t end = 0;
  while((end = s.find(delim, start)) != std::string::npos)
  {
    tokens.emplace_back(s.substr(start, end-start));
    start = end + delim.size(); 
  }
  tokens.emplace_back(s.substr(start));

  return tokens;
}

void solve(const std::vector<std::string>& rawInput)
{
  // std::cout << "Printing the input inside solution():" << std::endl;
  // for (auto line : rawInput) {
  //   std::cout << "Line: " << line << std::endl;
  // }

  std::string gamma;
  std::string not_gamma;
  for (size_t j = 0; j < rawInput[0].size(); ++j)
  {
    int count = 0;
    for (size_t i = 0; i < rawInput.size(); ++i)
    {
      count += (rawInput[i][j] == '1') - (rawInput[i][j] == '0');
    }
    std::cout << " - count: " << count << std::endl;
    gamma += '0' * (count < 0) + '1' * (count >= 0);
    not_gamma += '1' * (count < 0) + '0' * (count >= 0);
  }

  const int gamma_int = std::stoi(gamma, 0, 2);
  const int not_gamma_int = std::stoi(not_gamma, 0, 2);
  const int power = gamma_int * not_gamma_int;

  std::cout << "Power: " << power << std::endl;
}

void run(std::string_view filename)
{
  // Check if the file exists.
  const fs::path filepath { filename };
  if (!fs::exists(filepath))
  {
    std::cout << "File " << filepath.string() << " does not exist!" << std::endl;
    return;
  }

  std::cout << "Running the program using the input file: " << filepath.string() << std::endl;

  // Read the raw input and pass it in.
  std::vector<std::string> rawInput;

  std::ifstream file(filepath, std::ios::in);
  for(std::string line; std::getline(file, line); )
  {
    rawInput.emplace_back(line);
  }

  solve(rawInput);
}

int main(int argc, char** argv)
{
  // Parse arguments.
  const char* INPUT_FILENAME = { "input.txt" };
  std::string_view filename;
  if (argc < 2)
  {
    // std::cout << "Missing filename argument!" << std::endl;
    filename = INPUT_FILENAME;
  }
  else
  {
    filename = argv[1];
  }

  // Run the program.
  run(filename);

  return 0;
}